### DISCUSSION TITLE
brotli & zstandard support

### DISCUSSION THREAD

**Discussion by Do1e** [encode/httpx#3346] [General]
I strongly not recommend to continue using the current decoding strategy for Brotli & Zstandard.

According to my attempts, when the response compression method is br, `response.text` is displayed as garbled. At first I thought it was a problem of text encoding, which wasted a lot of time.

According to `httpx/_decoders.py`, I found that there are errors when the decoder initializes. But because of [pop](https://github.com/encode/httpx/blob/1bf1fc0ea8417c8e539914aeafdb48f15591ba75/httpx/_decoders.py#L368), nothing is done.

**Comment by lovelydinosaur** [comment]
I think I know what's going on there, tho could you demonstrate an example against a public URL to help us talk this through?

**Comment by Do1e** [comment]
An example on Archlinux:

```
> conda create -n httpx python=3.12
> conda activate httpx
> pip install "httpx[http2]"
> python test.py
```
![image](https://github.com/user-attachments/assets/794d7c56-ae1d-4af3-9ada-c5b84c4aa362)


```python
# test.py
import httpx

url = 'https://ip.011102.xyz/'
response = httpx.get(url, headers={'accept-encoding': 'zstd'})
print(response.headers.get('content-encoding'))
print(response.text)
```

If change `headers={'accept-encoding': 'zstd'}` to `headers={'accept-encoding': 'gzip'}` or install `zstandard`, the result will be correct.

If delete following lines in [_decoders.py](https://github.com/encode/httpx/blob/master/httpx/_decoders.py)
```python
if brotli is None:
    SUPPORTED_DECODERS.pop("br")  # pragma: no cover
if zstd is None:
    SUPPORTED_DECODERS.pop("zstd")  # pragma: no cover
```
It will report the error as expected.

![image](https://github.com/user-attachments/assets/31891bf0-ae4d-49eb-9766-66cab2ab8402)

**Reply by Do1e** [reply]
I realized the above porblems after trying different text encoding and carefully reading the document, which wasted a lot of time. I think it's better to give the error at first.
